File: Prototype/Assets/Scripts/MapGenerator/CellFenceGenerator.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using FenceEntity = CellFenceConfig.FenceEntity;

public class CellFenceGenerator
{
    /*
    class FenceSection
    {
        public Vector2 position;
        public Vector2 direction;
        public bool withEnd = false;
    }
    List<FenceSection> fenceSections;
    */

    class FenceObject
    {
        public Vector2 position;
        public Vector2 direction;
        public FenceEntity entity;
    }
    //List<FenceObject> fenceObjects;

    class Gate : System.IComparable<Gate>
    {
        public Vector2 position;
        public Vector2 direction;

        public int CompareTo(Gate other)
        {
            return Vector2.SignedAngle(Vector2.up, position).CompareTo(
                Vector2.SignedAngle(Vector2.up, other.position)
                );
        }
    }
    List<Gate> gates;

    class Segment
    {
        public Vector2 position;
        public Vector2 direction;
        public int fragmentCount;
        public float fragmentLength;

        public Vector2 endPosition { get { return position + direction * fragmentLength * fragmentCount; } }
    }
    class BrokenCurve
    {
        public List<Vector2> points;
        public List<Vector2> reversedPoints;

        public List<Segment> forawrdSegments = new List<Segment>();
        public List<Segment> backwardSegments = new List<Segment>();

        public int[] targetPointI = new int[] { 1 };
        public int[] targetReversedPointI = new int[] { 1 };

        public List<FenceObject> objects;
        public List<FenceObject> filler;

        public float HoleDistance()
        {
            Vector2 from = points[0];
            if (forawrdSegments.Count > 0)
                from = forawrdSegments[forawrdSegments.Count - 1].endPosition;

            Vector2 to = reversedPoints[0];
            if (backwardSegments.Count > 0)
                to = backwardSegments[backwardSegments.Count - 1].endPosition;

            return (from - to).magnitude;
        }
    }
    List<BrokenCurve> brokenCurves;

    CellFenceConfig config;
    ConvexPolygon polygon;

    public CellFenceGenerator(CellFenceConfig config, ConvexPolygon polygon)
    {
        this.config = config;
        this.polygon = polygon;
        gates = new List<Gate>();
    }

    public GameObject CreateGate(GameObject bridge, GameObject parent)
    {
        Vector3 offsetPosition = bridge.transform.position - parent.transform.position;
        Vector2 bridgePosition = new Vector2(
            offsetPosition.x,
            offsetPosition.z
            );
        Vector2 frontDirection = new Vector3(
            bridge.transform.forward.x,
            bridge.transform.forward.z
            ).normalized;
        Vector2 gateDirection = new Vector3(
            bridge.transform.right.x,
            bridge.transform.right.z
            ).normalized;

        if (bridgePosition.magnitude < (bridgePosition + frontDirection).magnitude)
        {
            frontDirection = -frontDirection;
            gateDirection = -gateDirection;
        }

        Vector2 gatePosition = bridgePosition + frontDirection * config.gateDistance;

        GameObject element = GameObject.Instantiate(config.gateEntity.gameObject);
        element.transform.parent = parent.transform;
        element.transform.localPosition = new Vector3(
            gatePosition.x,
            0.0f,
            gatePosition.y
            );
        element.transform.right = new Vector3(
            gateDirection.x,
            0.0f,
            gateDirection.y
            );

        gates.Add(new Gate() { position = gatePosition, direction = gateDirection });

        return element;
    }

    public void PrepareBrokenCurves()
    {
        brokenCurves = new List<BrokenCurve>();
        gates.Sort();

        List<Vector2> outline = new List<Vector2>();
        outline.AddRange(polygon.Points);
        outline.AddRange(polygon.Points); // make it overlap for easier indexing

        Gate fromGate, toGate;
        Vector2 startPoint, endPoint;
        for (int i=0; i<gates.Count; i++)
        {
            fromGate = gates[i];
            toGate = gates[i == gates.Count - 1 ? 0 : i + 1];

            startPoint = fromGate.position + fromGate.direction * config.gateEntity.length * 0.5f;
            endPoint = toGate.position - toGate.direction * config.gateEntity.length * 0.5f;

            int fromIndex = 0, toIndex = 0;
            for (int j=0; j<polygon.Points.Length; j++)
            {
                if ((startPoint - outline[j]).magnitude
                    < (startPoint - outline[fromIndex]).magnitude)
                    fromIndex = j;
                if ((endPoint - outline[j]).magnitude
                    < (endPoint - outline[toIndex]).magnitude)
                    toIndex = j;
            }

            if (toIndex < fromIndex)
                toIndex += polygon.Points.Length;

            if (toIndex <= 2)
            {
                fromIndex += polygon.Points.Length;
                toIndex += polygon.Points.Length;
            }

            
            if (Vector2.Dot(
                (outline[fromIndex + 1] - outline[fromIndex]).normalized,
                (startPoint - outline[fromIndex]).normalized
                ) > -0.2f)
                fromIndex++;
            if (Vector2.Dot(
                (outline[toIndex - 1] - outline[toIndex]).normalized,
                (endPoint - outline[toIndex]).normalized
                ) > -0.2f)
                toIndex--;

            if (Vector2.Dot(
                (outline[fromIndex + 1] - outline[fromIndex]).normalized,
                (startPoint - outline[fromIndex]).normalized
                ) > -0.2f)
                fromIndex++;
            if (Vector2.Dot(
                (outline[toIndex - 1] - outline[toIndex]).normalized,
                (endPoint - outline[toIndex]).normalized
                ) > -0.2f)
                toIndex--;


            List<Vector2> curve = new List<Vector2>();
            curve.Add(startPoint);
            for (int k = fromIndex; k <= toIndex; k++)
                curve.Add(outline[k]);
            curve.Add(endPoint);

            List<Vector2> reversedCurve = new List<Vector2>(curve);
            reversedCurve.Reverse();

            brokenCurves.Add(new BrokenCurve()
            {
                points = curve,
                reversedPoints = reversedCurve
            });
        }
    }

    public void CreateSections()
    {
        float lengthForNextSegment = config.fragmentEntity.length * config.fragmentCount + config.connectorEntity.length;
        float minHoleGap = lengthForNextSegment + config.connectorEntity.length;

        List<Vector2> targetPoints;
        List<Segment> targetSegments;
        int[] targetPointI;

        Vector2 nextPosition;
        Vector2 direction;
        Segment lastSegment;

        foreach (var curve in brokenCurves)
        {
            int i = 0;
            int created = 0;
            while (curve.HoleDistance() > minHoleGap || created < 2 )
            {
                if (i % 2 == 0)
                {
                    targetPoints = curve.points;
                    targetSegments = curve.forawrdSegments;
                    targetPointI = curve.targetPointI;
                }
                else
                {
                    targetPoints = curve.reversedPoints;
                    targetSegments = curve.backwardSegments;
                    targetPointI = curve.targetReversedPointI;
                }

                nextPosition = targetPoints[0];
                if (targetSegments.Count > 0)
                {
                    lastSegment = targetSegments[targetSegments.Count - 1];
                    nextPosition = lastSegment.position + lastSegment.direction * lengthForNextSegment;
                }

                direction = targetPoints[targetPointI[0]] - nextPosition;
                if (direction.magnitude >= lengthForNextSegment)
                {
                    targetSegments.Add(new Segment()
                    {
                        position = nextPosition,
                        direction = direction.normalized,
                        fragmentCount = config.fragmentCount,
                        fragmentLength = config.fragmentEntity.length
                    });
                    created++;
                }
                else if (targetPointI[0] < targetPoints.Count - 1)
                {
                    targetPointI[0]++;
                }

                i++;
            }
        }

    }

    public void TryFillGapsInSections()
    {
        Segment forward, backward;
        foreach (var curve in brokenCurves)
        {
            forward = curve.forawrdSegments[curve.forawrdSegments.Count - 1];
            backward = curve.backwardSegments[curve.backwardSegments.Count - 1];
            curve.filler = TryFillGap(forward, backward);
        }
    }
    List<FenceObject> TryFillGap(Segment forward, Segment backward)
    {
        List<FenceObject> filler = new List<FenceObject>();

        FenceEntity fragment = config.fragmentEntity;
        FenceEntity connector = config.connectorEntity;
        float conLen2 = connector.length * 2.0f;

        Vector2 forwardDirFrom = forward.direction;
        Vector2 forwardDirTo = (backward.position - forward.position).normalized;
        Vector2 backwarddDirFrom = backward.direction;
        Vector2 backwardDirTo = (forward.position - backward.position).normalized;

        for (float f = 1.0f; f >= 0.0f; f -= 0.2f)
        {
            forward.direction = Vector2.Lerp(
                    forwardDirFrom, forwardDirTo, f
                ).normalized;
            backward.direction = Vector2.Lerp(
                    backwarddDirFrom, backwardDirTo, f
                ).normalized;

            float gap = (forward.endPosition - backward.endPosition).magnitude;
            Vector2 dir = (backward.endPosition - forward.endPosition).normalized;

            // maybe just 1 connector?
            if (connector.InFillRange(gap))
            {
                filler.Add(new FenceObject()
                {
                    position = forward.endPosition + dir * gap * 0.5f,
                    direction = dir,
                    entity = connector
                });
                return filler;
            }

            // maybe 2 connectors + fragment?
            if (fragment.InFillRange(gap - conLen2))
            {
                float fragLen = gap - conLen2;
                filler.Add(new FenceObject()
                {
                    position = forward.endPosition + dir * connector.length * 0.5f,
                    direction = dir,
                    entity = connector
                });
                filler.Add(new FenceObject()
                {
                    position = forward.endPosition + dir * (connector.length + fragLen * 0.5f),
                    direction = dir,
                    entity = fragment
                });
                filler.Add(new FenceObject()
                {
                    position = forward.endPosition + dir * (connector.length * 1.5f + fragLen),
                    direction = dir,
                    entity = connector
                });
                return filler;
            }

            // maybe 2 connectors + 2 fragments?
            if (fragment.InFillRange(gap - conLen2, 2))
            {
                float fragsLen = gap - conLen2;
                float fragHalfLen = (fragsLen - fragment.length) * 0.5f;
                filler.Add(new FenceObject()
                {
                    position = forward.endPosition + dir * connector.length * 0.5f,
                    direction = dir,
                    entity = connector
                });
                filler.Add(new FenceObject()
                {
                    position = forward.endPosition + dir * (connector.length + fragHalfLen),
                    direction = dir,
                    entity = fragment
                });
                filler.Add(new FenceObject()
                {
                    position = forward.endPosition + dir * (connector.length + fragHalfLen + fragment.length),
                    direction = dir,
                    entity = fragment
                });
                filler.Add(new FenceObject()
                {
                    position = forward.endPosition + dir * (connector.length * 1.5f + fragsLen),
                    direction = dir,
                    entity = connector
                });
                return filler;
            }


            if (f < 0.00001f)
            {
                // couldn't find fill?
                // remove one fragment from one segment and try again
                if (backward.fragmentCount == 2)
                {
                    backward.fragmentCount = 1;
                    f = 1.2f;
                }
                // couldnt find then? try with three fragments instead
                else if (backward.fragmentCount == 1)
                {
                    backward.fragmentCount = 3;
                    f = 1.2f;
                }
                // couldnt find even then? reverse change and cancel finding :c
                else if (backward.fragmentCount == 3)
                {
                    // leave it "random"
                    if (Mathf.FloorToInt(backward.position.magnitude) % 2 == 0)
                    {
                        forward.fragmentCount = 2;
                        backward.fragmentCount = 3;
                    }
                    else
                    {
                        forward.fragmentCount = 3;
                        backward.fragmentCount = 2;
                    }
                    break;
                }
            }
        }

        return filler;
    }

    public void MakeHoles()
    {
        // TODO - make it better
        foreach (var curve in brokenCurves)
        {
            if (curve.filler != null && curve.filler.Count == 0)
            {
                curve.forawrdSegments[curve.forawrdSegments.Count - 1].fragmentCount = 0;
                curve.backwardSegments[curve.backwardSegments.Count - 1].fragmentCount = 0;
            }
        }
    }

    public void PrepareObjects()
    {
        foreach (var curve in brokenCurves)
        {
            curve.objects = new List<FenceObject>();
            PrepareObjectsForSection(curve.forawrdSegments, curve.objects, false);
            PrepareObjectsForSection(curve.backwardSegments, curve.objects, true);
            if (curve.filler != null)
                curve.objects.AddRange(curve.filler);
        }
    }
    void PrepareObjectsForSection(List<Segment> segments, List<FenceObject> target, bool reversed)
    {
        for (int i=0; i<segments.Count; i++)
        {
            if (i > 0)
            {
                Vector2 connectorFrom = segments[i - 1].endPosition;
                Vector2 connectorTo = segments[i].position;
                Vector2 connectorDirection = (connectorTo - connectorFrom).normalized;
                target.Add(new FenceObject()
                {
                    position = (connectorFrom + connectorTo) * 0.5f,
                    direction = reversed ? -connectorDirection : connectorDirection,
                    entity = config.connectorEntity
                });
            }

            float dist = 0.0f;
            for (int j=0; j<segments[i].fragmentCount; j++)
            {
                float eDist = dist + config.fragmentEntity.length * 0.5f;
                target.Add(new FenceObject()
                {
                    position = segments[i].position + segments[i].direction * eDist,
                    direction = reversed ? -segments[i].direction : segments[i].direction,
                    entity = config.fragmentEntity
                });
                dist += config.fragmentEntity.length;
            }
        }
    }

    public void BuildSections(GameObject parent)
    {
        GameObject element, container;
        for (int i = 0; i < brokenCurves.Count; i++)
        {
            container = new GameObject(i.ToString());
            container.transform.parent = parent.transform;
            container.transform.position = parent.transform.position;

            foreach (var o in brokenCurves[i].objects)
            {
                element = GameObject.Instantiate(o.entity.gameObject);
                element.transform.parent = container.transform;
                element.transform.localPosition = new Vector3(
                    o.position.x,
                    0.0f,
                    o.position.y
                    );
                element.transform.right = new Vector3(
                    o.direction.x,
                    0.0f,
                    o.direction.y
                    );
            }
        }   
    }

    /*
    public void PrepareOutlineSections()
    {
        fenceSections = new List<FenceSection>();

        List<Vector2> controlPoints = new List<Vector2>(polygon.Points);
        controlPoints.Add(controlPoints[0]);

        float sectionLength = config.GetLength();
        float sectionLengthWithEnd = config.GetLengthWithEnd();

        int targetPointI = 1;
        Vector2 nextPosition = controlPoints[0];
        Vector2 direction;
        while (targetPointI < controlPoints.Count)
        {
            direction = controlPoints[targetPointI] - nextPosition;
            if (direction.magnitude >= sectionLengthWithEnd)
            {
                fenceSections.Add(new FenceSection()
                {
                    position = nextPosition,
                    direction = direction.normalized
                });
                nextPosition += direction.normalized * sectionLength;
            }
            else
            {
                targetPointI++;
            }
        }

        fenceSections[fenceSections.Count - 1].withEnd = true;
    }

    public void PrepareObjects()
    {
        fenceObjects = new List<FenceObject>();
        var entites = new List<CellFenceConfig.FenceEntity>(config.sectionEntities);
        var entitesWithEnd = new List<CellFenceConfig.FenceEntity>(config.sectionEntities);
        entitesWithEnd.Add(config.endingEntity);

        Debug.Log(fenceSections.Count);
        foreach (FenceSection section in fenceSections)
        {
            float dist = 0.0f;
            foreach (var entity in section.withEnd ? entitesWithEnd : entites)
            {
                float eDist = dist + entity.length * 0.5f;
                fenceObjects.Add(new FenceObject()
                {
                    prefab = entity.gameObject,
                    position = section.position + section.direction * eDist,
                    direction = section.direction,
                    length = entity.length
                });
                dist += entity.length;
            }
        }
    }
    */
    /*
    public void Build(GameObject parent)
    {
        GameObject element;
        foreach (var o in fenceObjects)
        {
            element = GameObject.Instantiate(o.entity.gameObject);
            element.transform.parent = parent.transform;
            element.transform.localPosition = new Vector3(
                o.position.x,
                0.0f,
                o.position.y
                );
            element.transform.right = new Vector3(
                o.direction.x,
                0.0f,
                o.direction.y
                );
        }
    }
    */
}

File: Main/Game/Game/MapGenerator/CellFenceGenerator.cpp
#include <glm/gtx/vector_angle.hpp>
#include <glm/gtx/rotate_vector.hpp>
#include "CellFenceGenerator.h"
#include "HFEngine.h"
#include "ECS/Components.h"

GameObject CellFenceGenerator::CreateGate(GameObject bridge, GameObject parent, Transform& parentTransform)
{
    glm::vec3 offsetPosition =
        HFEngine::ECS.GetComponent<Transform>(bridge).GetWorldPosition() -
        parentTransform.GetWorldPosition();
        //HFEngine::ECS.GetComponent<Transform>(parent).GetWorldPosition();

    glm::vec2 bridgePosition = glm::vec2(
        offsetPosition.x,
        offsetPosition.z
        );
    glm::vec2 frontDirection = glm::normalize(glm::vec2(
        HFEngine::ECS.GetComponent<Transform>(bridge).GetWorldFront().x,
        HFEngine::ECS.GetComponent<Transform>(bridge).GetWorldFront().z
        ));
    glm::vec2 gateDirection = glm::normalize(glm::vec2(
        HFEngine::ECS.GetComponent<Transform>(bridge).GetWorldRight().x,
        HFEngine::ECS.GetComponent<Transform>(bridge).GetWorldRight().z
        ));

    if (glm::length(bridgePosition) < glm::length(bridgePosition + frontDirection))
    {
        frontDirection = -frontDirection;
        gateDirection = -gateDirection;
    }

    glm::vec2 gatePosition = bridgePosition + frontDirection * config.gateDistance;

    GameObject element = HFEngine::ECS.CreateGameObject(parent);
    /*
    element.transform.parent = parent.transform;
    element.transform.localPosition = new glm::vec3(
        gatePosition.x,
        0.0f,
        gatePosition.y
        );
    element.transform.right = new glm::vec3(
        gateDirection.x,
        0.0f,
        gateDirection.y
        );
        */

    auto gate = std::make_shared<Gate>();
    gate->position = gatePosition;
    gate->direction = gateDirection;
    gates.push_back(gate);

    return element;
}

void CellFenceGenerator::PrepareBrokenCurves()
{
    std::sort(gates.begin(), gates.end(), [](std::shared_ptr<Gate> a, std::shared_ptr<Gate> b) {
        float angleA = glm::orientedAngle({ 0,1 }, glm::normalize(a->position));
        float angleB = glm::orientedAngle({ 0,1 }, glm::normalize(b->position));
        return angleA < angleB;
        });
    //std::reverse(gates.begin(), gates.end());

    std::vector<glm::vec2> outline;
    outline.insert(outline.end(), polygon.Points.begin(), polygon.Points.end());
    outline.insert(outline.end(), polygon.Points.begin(), polygon.Points.end()); // make it overlap for easier indexing

    std::shared_ptr<Gate> fromGate, toGate;
    glm::vec2 startPoint, endPoint;
    for (int i = 0; i < gates.size(); i++)
    {
        fromGate = gates[i];
        toGate = gates[i == gates.size() - 1 ? 0 : i + 1];

        // repalced startPoint with endPoint due to different transform space
        startPoint = fromGate->position - fromGate->direction * config.gateEntity.length * 0.5f;
        endPoint = toGate->position + toGate->direction * config.gateEntity.length * 0.5f;

        int fromIndex = 0, toIndex = 0;
        for (int j = 0; j < polygon.Points.size(); j++)
        {
            if (glm::length(startPoint - outline[j])
                < glm::length(startPoint - outline[fromIndex]))
                fromIndex = j;
            if (glm::length(endPoint - outline[j])
                < glm::length(endPoint - outline[toIndex]))
                toIndex = j;
        }

        if (toIndex < fromIndex)
            toIndex += polygon.Points.size();

        if (toIndex <= 2)
        {
            fromIndex += polygon.Points.size();
            toIndex += polygon.Points.size();
        }


        if (glm::dot(
        glm::normalize(outline[fromIndex + 1] - outline[fromIndex]),
            glm::normalize(startPoint - outline[fromIndex])
            ) > -0.2f)
            fromIndex++;
        if (glm::dot(
            glm::normalize(outline[toIndex - 1] - outline[toIndex]),
            glm::normalize(endPoint - outline[toIndex])
            ) > -0.2f)
            toIndex--;

        if (glm::dot(
            glm::normalize(outline[fromIndex + 1] - outline[fromIndex]),
            glm::normalize(startPoint - outline[fromIndex])
            ) > -0.2f)
            fromIndex++;
        if (glm::dot(
            glm::normalize(outline[toIndex - 1] - outline[toIndex]),
            glm::normalize(endPoint - outline[toIndex])
            ) > -0.2f)
            toIndex--;


        std::vector<glm::vec2> curve;
        curve.push_back(startPoint);
        for (int k = fromIndex; k <= toIndex; k++)
            curve.push_back(outline[k]);
        curve.push_back(endPoint);

        std::vector<glm::vec2> reversedCurve;
        reversedCurve.insert(reversedCurve.end(), curve.begin(), curve.end());
        std::reverse(reversedCurve.begin(), reversedCurve.end());

        auto brokenCurve = std::make_shared<BrokenCurve>();
        brokenCurve->points = curve;
        brokenCurve->reversedPoints = reversedCurve;
        brokenCurves.push_back(brokenCurve);
    }
}

void CellFenceGenerator::CreateSections()
{
    float lengthForNextSegment = config.fragmentEntity.length * config.fragmentCount + config.connectorEntity.length;
    float minHoleGap = lengthForNextSegment + config.connectorEntity.length;

    std::vector<glm::vec2>* targetPoints;
    std::vector<std::shared_ptr<Segment>>* targetSegments;
    int* targetPointI;

    glm::vec2 nextPosition;
    glm::vec2 direction;
    std::shared_ptr<Segment> lastSegment;

    for (auto curve : brokenCurves)
    {
        int i = 0;
        int created = 0;
        float holeDistance;
        while ((holeDistance = curve->HoleDistance()) > minHoleGap || i < 2)
        {
            if (i % 2 == 0)
            {
                targetPoints = &curve->points;
                targetSegments = &curve->forawrdSegments;
                targetPointI = &curve->targetPointI;
            }
            else
            {
                targetPoints = &curve->reversedPoints;
                targetSegments = &curve->backwardSegments;
                targetPointI = &curve->targetReversedPointI;
            }

            nextPosition = targetPoints->at(0);
            if (targetSegments->size() > 0)
            {
                lastSegment = targetSegments->at(targetSegments->size() - 1);
                nextPosition = lastSegment->position + lastSegment->direction * lengthForNextSegment;
            }

            direction = targetPoints->at(*targetPointI) - nextPosition;
            if (glm::length(direction) >= lengthForNextSegment)
            {
                auto segment = std::make_shared<Segment>();
                segment->position = nextPosition;
                segment->direction = glm::normalize(direction);
                segment->fragmentCount = config.fragmentCount;
                segment->fragmentLength = config.fragmentEntity.length;
                targetSegments->push_back(segment);
                created++;
            }
            else if (*targetPointI < targetPoints->size() - 1)
            {
                *targetPointI = *targetPointI + 1;
            }

            i++;
        }
    }
}

void CellFenceGenerator::TryFillGapsInSections()
{
    for (int i=0; i<brokenCurves.size(); i++)
    {
        int fi = brokenCurves[i]->forawrdSegments.size() - 1;
        int bi = brokenCurves[i]->backwardSegments.size() - 1;
        brokenCurves[i]->filler = TryFillGap(
            brokenCurves[i]->forawrdSegments[fi],
            brokenCurves[i]->backwardSegments[bi]
            );
    }
}
std::vector<std::shared_ptr<CellFenceGenerator::FenceObject>>
    CellFenceGenerator::TryFillGap(
        std::shared_ptr<CellFenceGenerator::Segment> forward,
        std::shared_ptr<CellFenceGenerator::Segment> backward
        )
{
    std::vector<std::shared_ptr<CellFenceGenerator::FenceObject>> filler;

    CellFenceEntity* fragment = &config.fragmentEntity;
    CellFenceEntity* connector = &config.connectorEntity;
    float conLen2 = connector->length * 2.0f;

    glm::vec2 forwardDirFrom = forward->direction;
    glm::vec2 forwardDirTo = glm::normalize(backward->position - forward->position);
    glm::vec2 backwarddDirFrom = backward->direction;
    glm::vec2 backwardDirTo = glm::normalize(forward->position - backward->position);

    for (float f = 1.0f; f >= 0.0f; f -= 0.2f)
    {
        forward->direction = glm::normalize(glm::mix(
            forwardDirFrom, forwardDirTo, f
            ));
        backward->direction = glm::normalize(glm::mix(
            backwarddDirFrom, backwardDirTo, f
            ));

        float gap = glm::length(forward->endPosition() - backward->endPosition());
        glm::vec2 dir = glm::normalize(backward->endPosition() - forward->endPosition());

        // maybe just 1 connector?
        if (connector->InFillRange(gap))
        {
            filler.push_back(std::make_shared<FenceObject>(
                forward->endPosition() + dir * gap * 0.5f,
                dir,
                connector
                ));
            return filler;
        }

        // maybe 2 connectors + fragment?
        if (fragment->InFillRange(gap - conLen2))
        {
            float fragLen = gap - conLen2;
            filler.push_back(std::make_shared<FenceObject>(
                    forward->endPosition() + dir * connector->length * 0.5f,
                    dir,
                    connector
                ));
            filler.push_back(std::make_shared<FenceObject>(
                    forward->endPosition() + dir * (connector->length + fragLen * 0.5f),
                    dir,
                    fragment
                ));
            filler.push_back(std::make_shared<FenceObject>(
                    forward->endPosition() + dir * (connector->length * 1.5f + fragLen),
                    dir,
                    connector
                ));
            return filler;
        }

        // maybe 2 connectors + 2 fragments?
        if (fragment->InFillRange(gap - conLen2, 2))
        {
            float fragsLen = gap - conLen2;
            float fragHalfLen = (fragsLen - fragment->length) * 0.5f;
            filler.push_back(std::make_shared<FenceObject>(
                    forward->endPosition() + dir * connector->length * 0.5f,
                    dir,
                    connector
                ));
            filler.push_back(std::make_shared<FenceObject>(
                    forward->endPosition() + dir * (connector->length + fragHalfLen),
                    dir,
                    fragment
                ));
            filler.push_back(std::make_shared<FenceObject>(
                    forward->endPosition() + dir * (connector->length + fragHalfLen + fragment->length),
                    dir,
                    fragment
                ));
            filler.push_back(std::make_shared<FenceObject>(
                    forward->endPosition() + dir * (connector->length * 1.5f + fragsLen),
                    dir,
                    connector
                ));
            return filler;
        }


        if (f < 0.00001f)
        {
            // couldn't find fill?
            // remove one fragment from one segment and try again
            if (backward->fragmentCount == 2)
            {
                backward->fragmentCount = 1;
                f = 1.2f;
            }
            // couldnt find then? try with three fragments instead
            else if (backward->fragmentCount == 1)
            {
                backward->fragmentCount = 3;
                f = 1.2f;
            }
            // couldnt find even then? reverse change and cancel finding :c
            else if (backward->fragmentCount == 3)
            {
                // leave it "random"
                if (((int)(glm::length(backward->position))) % 2 == 0)
                {
                    forward->fragmentCount = 2;
                    backward->fragmentCount = 3;
                }
                else
                {
                    forward->fragmentCount = 3;
                    backward->fragmentCount = 2;
                }
                break;
            }
        }
    }

    return filler;
}

void CellFenceGenerator::MakeHoles()
{
    // TODO - make it better
    for (auto curve : brokenCurves)
    {
        if (curve->filler.size() == 0)
        {
            curve->forawrdSegments[curve->forawrdSegments.size() - 1]->fragmentCount = 0;
            curve->backwardSegments[curve->backwardSegments.size() - 1]->fragmentCount = 0;
        }
    }
}

void CellFenceGenerator::PrepareObjects()
{
    for (auto curve : brokenCurves)
    {
        curve->objects.clear();
        PrepareObjectsForSection(curve->forawrdSegments, curve->objects, false);
        PrepareObjectsForSection(curve->backwardSegments, curve->objects, true);
        if (curve->filler.size() == 0)
            curve->objects.insert(curve->objects.end(), curve->filler.begin(), curve->filler.end());
    }
}
void CellFenceGenerator::PrepareObjectsForSection(std::vector<std::shared_ptr<Segment>>& segments, std::vector<std::shared_ptr<FenceObject>>& target, bool reversed)
{
    for (int i = 0; i < segments.size(); i++)
    {
        if (i > 0)
        {
            glm::vec2 connectorFrom = segments[i - 1]->endPosition();
            glm::vec2 connectorTo = segments[i]->position;
            glm::vec2 connectorDirection = glm::normalize(connectorTo - connectorFrom);
            target.push_back(std::make_shared<FenceObject>(
                    (connectorFrom + connectorTo) * 0.5f,
                    reversed ? -connectorDirection : connectorDirection,
                    &config.connectorEntity
                ));
        }

        float dist = 0.0f;
        for (int j = 0; j < segments[i]->fragmentCount; j++)
        {
            float eDist = dist + config.fragmentEntity.length * 0.5f;
            target.push_back(std::make_shared<FenceObject>(
                    segments[i]->position + segments[i]->direction * eDist,
                    reversed ? -segments[i]->direction : segments[i]->direction,
                    &config.fragmentEntity
                ));
            dist += config.fragmentEntity.length;
        }
    }
}

void CellFenceGenerator::BuildSections(GameObject parent)
{
    //if (parent > 400) return;
    GameObject element, container;
    for (int i = 0; i < brokenCurves.size(); i++)
    {
        container = HFEngine::ECS.CreateGameObject(parent, std::to_string(i));

        for (auto o : brokenCurves[i]->objects)
        {
            float rotation = rad2deg(glm::atan(o->direction.x, o->direction.y)) + 90; // + 90 for right
            element = o->entity->prefab->Instantiate(
                parent,
                { o->position.x, 0.0f, o->position.y },
                { 0.0f, rotation, 0.0f }
               );
        }
    }
}
